File: TexasHoldem/TexasHoldem/Models/HandEvaluator.swift
import Foundation

enum HandCategory: Int, Comparable, Codable {
    case highCard = 0
    case pair
    case twoPair
    case threeOfAKind
    case straight
    case flush
    case fullHouse
    case fourOfAKind
    case straightFlush

    static func < (lhs: HandCategory, rhs: HandCategory) -> Bool { lhs.rawValue < rhs.rawValue }

    var displayName: String {
        switch self {
        case .highCard: return "High Card"
        case .pair: return "Pair"
        case .twoPair: return "Two Pair"
        case .threeOfAKind: return "Three of a Kind"
        case .straight: return "Straight"
        case .flush: return "Flush"
        case .fullHouse: return "Full House"
        case .fourOfAKind: return "Four of a Kind"
        case .straightFlush: return "Straight Flush"
        }
    }
}

/// Ranks a 5-card hand. `tiebreakers` is ordered most-significant first,
/// e.g. for a full house [tripsRank, pairRank]; for high card, the five
/// ranks in descending order.
struct HandRank: Comparable, Codable {
    let category: HandCategory
    let tiebreakers: [Int]

    static func < (lhs: HandRank, rhs: HandRank) -> Bool {
        if lhs.category != rhs.category { return lhs.category < rhs.category }
        for (a, b) in zip(lhs.tiebreakers, rhs.tiebreakers) where a != b {
            return a < b
        }
        return false
    }

    static func == (lhs: HandRank, rhs: HandRank) -> Bool {
        lhs.category == rhs.category && lhs.tiebreakers == rhs.tiebreakers
    }
}

enum HandEvaluator {
    /// Evaluates the best possible 5-card hand out of 5-7 given cards
    /// (typically 2 hole cards + up to 5 community cards).
    static func bestHand(from cards: [Card]) -> HandRank {
        precondition(cards.count >= 5, "Need at least 5 cards to evaluate a hand")
        var best: HandRank?
        for combo in combinations(cards, choose: 5) {
            let rank = evaluateFive(combo)
            if best == nil || rank > best! { best = rank }
        }
        return best!
    }

    private static func evaluateFive(_ cards: [Card]) -> HandRank {
        let ranksDescending = cards.map { $0.rank.rawValue }.sorted(by: >)
        let isFlush = Set(cards.map { $0.suit }).count == 1

        let straightHigh = straightHighCard(ranksDescending)
        let isStraight = straightHigh != nil

        var countsByRank: [Int: Int] = [:]
        for r in ranksDescending { countsByRank[r, default: 0] += 1 }
        // Groups sorted by (count desc, rank desc)
        let groups = countsByRank.sorted { lhs, rhs in
            if lhs.value != rhs.value { return lhs.value > rhs.value }
            return lhs.key > rhs.key
        }
        let counts = groups.map { $0.value }

        if isStraight && isFlush {
            return HandRank(category: .straightFlush, tiebreakers: [straightHigh!])
        }
        if counts == [4, 1] {
            return HandRank(category: .fourOfAKind, tiebreakers: groups.map { $0.key })
        }
        if counts == [3, 2] {
            return HandRank(category: .fullHouse, tiebreakers: groups.map { $0.key })
        }
        if isFlush {
            return HandRank(category: .flush, tiebreakers: ranksDescending)
        }
        if isStraight {
            return HandRank(category: .straight, tiebreakers: [straightHigh!])
        }
        if counts == [3, 1, 1] {
            return HandRank(category: .threeOfAKind, tiebreakers: groups.map { $0.key })
        }
        if counts == [2, 2, 1] {
            return HandRank(category: .twoPair, tiebreakers: groups.map { $0.key })
        }
        if counts == [2, 1, 1, 1] {
            return HandRank(category: .pair, tiebreakers: groups.map { $0.key })
        }
        return HandRank(category: .highCard, tiebreakers: ranksDescending)
    }

    /// Returns the high card of the straight if the given descending, unique-checked
    /// rank list forms a straight (handles wheel A-2-3-4-5), else nil.
    private static func straightHighCard(_ ranksDescending: [Int]) -> Int? {
        let unique = Array(Set(ranksDescending)).sorted(by: >)
        guard unique.count == 5 else { return nil }
        if unique[0] - unique[4] == 4 { return unique[0] }
        // Wheel: A,5,4,3,2
        if unique == [14, 5, 4, 3, 2] { return 5 }
        return nil
    }

    private static func combinations<T>(_ array: [T], choose k: Int) -> [[T]] {
        guard k <= array.count else { return [] }
        if k == 0 { return [[]] }
        if k == array.count { return [array] }
        var result: [[T]] = []
        func helper(_ start: Int, _ chosen: [T]) {
            if chosen.count == k {
                result.append(chosen)
                return
            }
            guard start < array.count else { return }
            for i in start..<array.count {
                helper(i + 1, chosen + [array[i]])
            }
        }
        helper(0, [])
        return result
    }
}

File: poker-app/server/poker/handEvaluator.js
// Evaluates the best 5-card poker hand out of any N (>=5) cards.
// Returns { category, tiebreak, best5 } where higher category/tiebreak wins.
// category: 8 straight flush, 7 quads, 6 full house, 5 flush, 4 straight,
//           3 trips, 2 two pair, 1 pair, 0 high card.

const CATEGORY_NAMES = [
  'High Card', 'Pair', 'Two Pair', 'Three of a Kind', 'Straight',
  'Flush', 'Full House', 'Four of a Kind', 'Straight Flush',
];

function combinations(arr, k) {
  const results = [];
  const combo = [];
  function recurse(start) {
    if (combo.length === k) {
      results.push(combo.slice());
      return;
    }
    for (let i = start; i < arr.length; i++) {
      combo.push(arr[i]);
      recurse(i + 1);
      combo.pop();
    }
  }
  recurse(0);
  return results;
}

function evaluate5(cards) {
  const values = cards.map((c) => c.value).sort((a, b) => b - a);
  const suits = cards.map((c) => c.suit);
  const isFlush = suits.every((s) => s === suits[0]);

  const counts = new Map();
  for (const v of values) counts.set(v, (counts.get(v) || 0) + 1);
  const groups = [...counts.entries()].sort((a, b) => {
    if (b[1] !== a[1]) return b[1] - a[1];
    return b[0] - a[0];
  });

  const uniqueDesc = [...new Set(values)];
  let straightHigh = null;
  if (uniqueDesc.length === 5) {
    if (uniqueDesc[0] - uniqueDesc[4] === 4) {
      straightHigh = uniqueDesc[0];
    } else if (uniqueDesc.join(',') === '14,5,4,3,2') {
      straightHigh = 5; // wheel: A-2-3-4-5
    }
  }

  if (straightHigh && isFlush) {
    return { category: 8, tiebreak: [straightHigh], name: 'Straight Flush' };
  }
  if (groups[0][1] === 4) {
    const kicker = groups.find((g) => g[1] === 1)[0];
    return { category: 7, tiebreak: [groups[0][0], kicker], name: 'Four of a Kind' };
  }
  if (groups[0][1] === 3 && groups[1] && groups[1][1] >= 2) {
    return { category: 6, tiebreak: [groups[0][0], groups[1][0]], name: 'Full House' };
  }
  if (isFlush) {
    return { category: 5, tiebreak: values, name: 'Flush' };
  }
  if (straightHigh) {
    return { category: 4, tiebreak: [straightHigh], name: 'Straight' };
  }
  if (groups[0][1] === 3) {
    const kickers = groups.filter((g) => g[1] === 1).map((g) => g[0]).sort((a, b) => b - a);
    return { category: 3, tiebreak: [groups[0][0], ...kickers], name: 'Three of a Kind' };
  }
  if (groups[0][1] === 2 && groups[1] && groups[1][1] === 2) {
    const pairs = [groups[0][0], groups[1][0]].sort((a, b) => b - a);
    const kicker = groups.find((g) => g[1] === 1)[0];
    return { category: 2, tiebreak: [...pairs, kicker], name: 'Two Pair' };
  }
  if (groups[0][1] === 2) {
    const kickers = groups.filter((g) => g[1] === 1).map((g) => g[0]).sort((a, b) => b - a);
    return { category: 1, tiebreak: [groups[0][0], ...kickers], name: 'Pair' };
  }
  return { category: 0, tiebreak: values, name: 'High Card' };
}

function compareResults(a, b) {
  if (a.category !== b.category) return a.category - b.category;
  for (let i = 0; i < Math.max(a.tiebreak.length, b.tiebreak.length); i++) {
    const av = a.tiebreak[i] || 0;
    const bv = b.tiebreak[i] || 0;
    if (av !== bv) return av - bv;
  }
  return 0;
}

// cards: array of >=5 card objects {rank, suit, value}
function evaluateBest(cards) {
  if (cards.length === 5) return { ...evaluate5(cards), cards };
  let best = null;
  for (const combo of combinations(cards, 5)) {
    const result = evaluate5(combo);
    if (!best || compareResults(result, best) > 0) {
      best = { ...result, cards: combo };
    }
  }
  return best;
}

module.exports = { evaluateBest, compareResults, CATEGORY_NAMES };
